libherokubuildpack/src/inventory.rs

```diff
@@ -1,3 +1,76 @@
+//! # Inventory
+//!
+//! Many buildpacks need to provide artifacts from different URLs. A helpful pattern
+//! is to provide a list of artifacts in a TOML file, which can be parsed and used by
+//! the buildpack to download the correct artifact. For example, a Ruby buildpack
+//! might need to download pre-compiled Ruby binaries hosted on S3.
+//!
+//! This module can be used to produce and consume such an inventory file.
+//!
+//! ## Features
+//!
+//! - Version lookup and comparison: To implement the inventory, you'll need to define how
+//!   versions are compared. This allows the inventory code to find an appropriate artifact
+//!   based on whatever custom version logic you need. If you don't need custom logic, you can
+//!   use the included `inventory-semver` feature.
+//! - Architecture aware: Beyond version specifiers, buildpack authors may need to provide different
+//!   artifacts for different computer architectures such as ARM64 or AMD64. The inventory encodes
+//!   this information which is used to select the correct artifact.
+//! - Checksum validation: In addition to knowing the URL of an artifact, buildp authors
+//!   want to be confident that the artifact they download is the correct one. To accomplish this
+//!   the inventory contains a checksum of the download and can be used to validate the download
+//!   has not been modified or tampered with. To use sha256 or sha512 checksums out of the box,
+//!   enable the `inventory-sha2` feature
+//! - Extensible with metadata: The default inventory format covers a lot of common use cases,
+//!   but if you need more, you can extend it by adding custom metadata to each artifact.
+//!
+//! ## Example consumer
+//!
+//! This example uses the `inventory-semver` and `inventory-sha2` features to parse an existing
+//! inventory file, compare versions via semver logic:
+//!
+//! ```no_run,rust
+//! use libherokubuildpack::inventory::{artifact::{Os, Arch}, Inventory, checksum::Checksum};
+//! use semver::{Version, VersionReq};
+//! use libherokubuildpack::download::download_file;
+//! use std::path::Path;
+//! use std::str::FromStr;
+//!
+//! #[cfg(feature = "inventory-sha2")]
+//! #[cfg(feature = "inventory-semver")]
+//! use sha2::Sha256;
+//!
+//! let inventory: Inventory<Version, Sha256, Option<()>> =
+//! std::fs::read_to_string("inventory.toml")
+//!     .unwrap()
+//!     .parse()
+//!     .unwrap();
+//! let requirement = VersionReq::parse("= 1.0.0").unwrap();
+//! if let Some(artifact) = inventory.resolve(Os::Linux, Arch::Amd64, &requirement) {
+//!     // Downloading the artifact
+//!     println!("Installing {requirement:?} from {}", artifact.url);
+//!     let path = Path::new("path/to/binary");
+//!     download_file(&artifact.url, &path)
+//!        .unwrap();
+//!
+//!     // Validating the checksum
+//!     Checksum::<Sha256>::from_str(&std::fs::read_to_string(&path).unwrap())
```

Comment:
The `FromStr` implementation for `Checksum` doesn't calculate the checksum for a file like this. It just parses a string containing the checksum value and name, and validating the format against the `Digest` implementation, e.g.:

    Checksum::<Sha256>::from_str(
        "sha256:2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
    )

Reply:
https://github.com/heroku/libcnb.rs/pull/864/commits/28204a087982c968e23e6b2b8c0e6b2b4ce26ed2 addresses this

Reply:
Good catch. I was wanting to show a few things that were not intuitive (clearly, since my example is wrong). I wanted to show:

- Generating a SHA from a file on disk
- Using that value to compare to the digest from the inventory
- Printing a human readable version of the sha

I will add a line to the example that shows `hex::encode` for printing.